### ISSUE TITLE
add context support in dataloader

### ISSUE THREAD

**Issue by oonxt** [async-graphql/async-graphql#1002]
## add context support in dataloader

inspired by hasura:

https://hasura.io/docs/latest/queries/postgres/query-filters/

## Code example (if possible)

```graphql
{
   users {
    orders(where: {xxx} limit: 10) {
      id paid_at
    }
  }
}
```

**Comment by onx2** [comment]
Are you looking to extend `load`?

```rust 
/// Trait for batch loading.
#[async_trait::async_trait]
pub trait Loader<K: Send + Sync + Hash + Eq + Clone + 'static>: Send + Sync + 'static {
    /// type of value.
    type Value: Send + Sync + Clone + 'static;

    /// Type of error.
    type Error: Send + Clone + 'static;

    /// Load the data set specified by the `keys`.
    async fn load(&self, keys: &[K]) -> Result<HashMap<K, Self::Value>, Self::Error>;
}
```

**Comment by alucryd** [comment]
Would be very nice indeed as it would allow for sharing a db pool across the board, the current implementation requires the pool to belong to the loader, making it impossible to have it in the context for queries, mutations, etc...

**Comment by vstepanyuk** [comment]
I would like also to forward some request headers to the upstream service from dataloader, having ctx: &Context<'_> would be probably enough

**Comment by humb1t** [comment]
Agree, would be great and sufficient for most cases to add `ctx` as a parameter into loader.
If maintainers agree with this design, we can create a PR with initial implementation.

**Comment by humb1t** [comment]
@Miaxos what do you think? It seems like an idiomatic way — async-graphql passes context to queries and mutations, we have use cases to pass it into loaders also.
The main problems I see are:
- Data loaders are stored in Context themselves;
- Data loaders are pretty lightweight right now, with only keys to load being stored + cache if used.

So maybe the right design choice should be different from just sending context directly to `load` methods of `Loader` trait. I will try to research codebase and create some examples with alternative approaches.

**Comment by Miaxos** [comment]
> Agree, would be great and sufficient for most cases to add ctx as a parameter into loader.

I'm not too sure about this:

The `Context` describe "where" you are when you process something, the purpose of dataloader is to delay the execution of an async function to batch multiples keys within, if you have your dataloader called from multiple places, which `Context do you use`?

---

I had this use case before when I wanted to share a db pool across the whole query execution depending on the query; what I did was storing the `PgPool` at the beginning of the request inside a struct where I implemented a `Dataloader` and depending on the need, I had one long transaction running until the end of the request or I used the inner pool of my dataloader.

At the end the dataloader impl in `async-graphql` is only an example of the pattern, you can use [dataloader-rs](https://github.com/cksac/dataloader-rs) for instance or even your custom trait/implementation that fit your needs.

**Comment by Miaxos** [comment]
> Example Scenario of Context

I'm sure there are examples, but I believe this shouldn't be the `Context` as it got some meaning over the "where" from where it's executed, it could change every time the dataloader is called. What you could put would be more something like the `ExtensionContext`.

> Instead of passing with each dataloader key - required headers to the dataloader, I would read them from the context in load function.

If the purpose is to not read them each time you call the dataloader, would it work to store them inside the dataloader struct at the beginning of the request and when you `load` you would be able to read them from the dataloader struct?

**Comment by Miaxos** [comment]
I was mentioning this to put data at the beginning of the request:

https://async-graphql.github.io/async-graphql/en/context.html#request-data

**Comment by Miaxos** [comment]
I would suggest to have an domain type `AuthContext` which would be created from the headers for instance instead of just the headers but this is the idea yes.

**Comment by humb1t** [comment]
Thank you, Anthony 💜

**Comment by ziimakc** [comment]
I wanted to heavenly use dataloaders, but creating multiple pools with active db connections maybe problematic :(

**Comment by xamgore** [comment]
@ziimakc, why?

```rust
#[derive(Debug, Clone)]
pub struct MyLoader {
    pub db: Arc<Pool>,
}

impl async_graphql::dataloader::Loader<..> for MyLoader {
    async fn load(&self, keys: &[..]) { ... }
}

fn main() {
    // let pool = ...;
    Schema::build(Query::default(), Mutation::default(), EmptySubscription)
        .data(DataLoader::new(Loader::new(pool.clone()), tokio::spawn))
}
```
